### PR TITLE
[FIX] mail: emoji data for search and display in other languages

### PR DESCRIPTION
**Pull request by rapr-odoo**
**Current behavior before PR:**

Emoji data are all in English. Unable to use in other languages.

**Desired behavior after PR is merged:**

Emoji names and keywords are translatable for search and display purposes.
Emoji category names are translateable for display.

Task-3043336

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/mail/static/src/models/emoji_registry.js [hunk 1]
```diff
@@ -33,8 +33,8 @@ registerModel({
                 codepoints: emojiData.codepoints,
                 shortcodes: emojiData.shortcodes,
                 emoticons: emojiData.emoticons,
-                name: emojiData.name,
-                keywords: emojiData.keywords,
+                name: this.env._t(emojiData.name),
+                keywords: emojiData.keywords.map(keyword => this.env._t(keyword)),
```
Comment (alexkuhn) [inline addons/mail/static/src/models/emoji_registry.js:37]: `_t()` requires the string to be in plain. It doesn't work with variables.
So the wrapping of `_t` must be done much earlier.

**Review by alexkuhn — CHANGES_REQUESTED** [review]

File: addons/mail/static/src/models/emoji_registry.js (hunk 1 above)
Comment (rapr-odoo) [reply addons/mail/static/src/models/emoji_registry.js]: I did test it and it was working. As even if they are keywords they hold String values only. 

I also got into an issue where if we replace the translated keywords with the original, we won't be able to make any search based on the original (ex. 'search' and it is also a common search parameter). So, I added some fixup for it. So, please take a look at that. Thanks!

Comment (alexkuhn) [reply addons/mail/static/src/models/emoji_registry.js]: It's working if we have the entries in `POT` file.
Problem here is that the strings are not automatically registered in the `POT` file.
The solution to this problem is to wrap `_t()` where the string are in plain, which is in each occurrences of translatable strings in `emoji_data.js`.

**Comment by rapr-odoo** [comment]
@aka-odoo Changes Done!

**Comment by rapr-odoo** [comment]
@alexkuhn Changes Done!

**Review by alexkuhn — CHANGES_REQUESTED** [review]
Instead of concatenation (`+`), you could use interpolation (`${}`).

**Comment by alexkuhn** [comment]
@mart-e Hello,
Emoji picker texts are not properly flagged as translatable.
This commit fixes the issue by flagging all of them, including keywords used to search emojis.
There are a lot of new entries in `mail.pot` as a result.
Is it okay as a change in stable version?
Thanks!

**Comment by mart-e** [comment]
>  Is it okay as a change in stable version?

Adding new terms is ok, changing is still possible in 16 as we just released.

>  There are a lot of new entries in mail.pot as a result.

Indeed it seems to be quite a lot. Is everything needed? We triple the number of translated strings in `mail` which was already quite high, from 1300 to 4600 strings (while it's 160 strings per module on average).
With this change, mail would become the biggest source of translated terms (more than `base` with 4300 strings). That will be a lot of work for translators.
Are emoji keywords translated in different messaging apps (whatsapp, discord,...) ?

**Comment by alexkuhn** [comment]
> Are emoji keywords translated in different messaging apps (whatsapp, discord,...) ?

I don't use any messaging app in other language than English, but apparently they do, which makes actually sense!
@LNAodoo asked many time for this feature, so I suppose this is really useful.

**Comment by rapr-odoo** [comment]
> Instead of concatenation (`+`), you could use interpolation (`${}`).

**Comment by rapr-odoo** [comment]
> Instead of concatenation (`+`), you could use interpolation (`${}`).

Initially, I did try to do it using interpolation. But, the `Translator Tool` was picking the whole emoji data as a single string while creating `mail.pot` file, including the terms wrapped in `${_t()}`.

**Comment by LNAodoo** [comment]
> 

Hello there, 
Answering here since I've been pinged :) 
I can't say much regarding the technical implementation and its consequences but I confirm that apps that allow people to search for emojis indeed usually translate them (I had a quick look at Facebook  & my mobile phone keyboard, for example).
Otherwise displaying a search that retrieves nothing if you search for "sourire" would be very weird for users.
Hope it helps,
Cheers!

**Comment by mart-e** [comment]
Well then, if it's what we want, let's go for it.
However, it seems there is some unecessary ones, like `1.30`, `11:00`, `1234` or `AB` :man_shrugging:

**Comment by alexkuhn** [comment]
> > Instead of concatenation (`+`), you could use interpolation (`${}`).
> 
> Initially, I did try to do it using interpolation. But, the `Translator Tool` was picking the whole emoji data as a single string while creating `mail.pot` file, including the terms wrapped in `${_t()}`.

Ok, not a big deal so ok to keep concatenation then.
Make a pass on all keywords and remove the `_t()` by hand where it's not useful. Such as all clock time `00`, `12`, `12:00`, or "*" (for star/sparkle emoji), or symbols-specific texts like  `abcd`, `1234`, `〒♪&%`.

**Comment by rapr-odoo** [comment]
Changes Done!

**Comment by alexkuhn** [comment]
> Make a pass on all keywords and remove the `_t()` by hand where it's not useful. Such as all clock time `00`, `12`, `12:00`, [...], or symbols-specific texts like [...] `1234` [...].

Actually... this might not be a good idea, as some languages have different characters for numeral. For example in Japanese it's １２３４ instead of 1234... And I just learned it's the same with sparkle * that is ＊...

So actually should turn translation for them, if we want to properly cover such languages. What do you think @mart-e ?

**Comment by mart-e** [comment]
> So actually should turn translation for them, if we want to properly cover such languages. What do you think @mart-e ?

I expect translators will not understand what those are about and that `A` will be translated `Un` in French but that's not a blocking point either. So ok :)

**Comment by rapr-odoo** [comment]
Changes Done!

**Comment by alexkuhn** [comment]
Here is list of all keywords that are not translated at the moment of writing:

- rofl (`:rofl:`)
- ILY (`:love-you_gesture:`)
- 12:00 (`:twelve_o’clock:`)
    - 12:30/1:00 etc.
- 8 (`:pool_8_ball:`)
- FM (`:radio:`)
- PC/pc (`:laptop:`)
- CD (`:optical_disk:`)
- dvd/DVD (`:dvd:`)
- VHS/vhs (`:videocassette:`)
- `+` (`:plus:`)
- `-` (`:minus:`)
- ÷ (`:divide:`)
- !/!! (`:double_exclamation_mark:`)
- !/!?/? (`:exclamation_question_mark:`)
- ? (`:red_question_mark:`/`:white_question_mark:`)
- ! (`:white_exclamation_mark:`/`:red_exclamation_mark:`)
- o (`:hollow_red_circle:`)
- ✓ (`:check_mark_button:`/`:check_box_with_check:`/`:check_mark:`)
- C (`:copyright:`)
- R (`:registered:`)
- TM (`:trade_mark:`)
- ABCD (`:input_latin_uppercase:`)
- abcd (`:input_latin_lowercase:`)
- 〒♪&% (`:input_symbols:`)
- abc (`:input_latin_letters:`)
- A (`:A_button_(blood_type):`)
     - AB/B/CL etc.
- i (`:information:`)
- ID (`:ID_button:`)
- M (`:circle_M:`)
- NG (`:NG_button:`)
- O (`:O_button_(blood_type):`)
- P (`:P_button:`)
- SOS (`:SOS_button:`)
- UP! (`:UP!_button:`)
- VS (`:VS_button:`)
- ココ (`:Japanese_"here"_button:`)
    - サ/月/etc.
    
Some should be flagged for translation, like "rofl" (`:rofl:`) that should become "mdr" in French, or "o" (`:hollow_red_circle`) that should become "◯" in Japanese (FYI in some contexts in Japanese, the opposite is `:cross_mark:`, which has "×" properly flagged for translation to become "✗").

As for latin characters, it's usually latin, but it fails with some IME.
For example, for "A":
```
'A' === 'Α' === 'А'; // false in all combination
```
latin, greek, cyrillic. These are very common character input for some languages, so they must be translated too.

In the end, very few cases to not mark for translation, like the Japanese character symbols.

**Comment by alexkuhn** [comment]
So to sum up, I think it's best to flag all keywords for traduction, as there are almost no keywords that shouldn't need translation.

**Comment by rapr-odoo** [comment]
Changes Done!

**Comment by alexkuhn** [comment]
👆use `_lt` instead of `_t` (just in case this is loaded when translation are not yet loaded), and also use the translation in `l10n/translation` rather than `web.core` (deprecated).

**Pull request closed without merging**